File: data-structures/linked-list/linked-list.go
package linkedlist

import "fmt"

type Node struct {
	data string
	next *Node
}

func NewLinkedList(initialValue string) *Node {
	head := &Node{
		data: initialValue,
		next: nil,
	}

	return head
}

func (node *Node) AddNode(value string) {
	isTail := node.next == nil
	if isTail {
		node.next = &Node{
			data: value,
			next: nil,
		}
	} else {
		newNode := &Node{
			data: value,
			next: node.next,
		}

		node.next = newNode
	}
}

func (node *Node) Print() {
	for node != nil {
		fmt.Println(node.data)
		node = node.next
	}
}

File: c/linked-list/linked-list.c
#include <stdio.h>
#include <stdlib.h>
#include "linked-list.h"

// TODO: head of linked list already with a value
Node* LinkedList(int v) {
	Node *head = malloc(sizeof(Node));
	head->value = v;
	head->next = NULL;
	return head;
}

void print(Node *head) {
	Node *tmp = head;
	while(tmp != NULL) {
		printf("%d\n", tmp->value);
		tmp = tmp->next;
	};
}

Node* search(Node *head, int n) {
	Node *temp = head;
	while(temp != NULL && temp->value != n) {
		temp = temp->next;	
	}
	return temp;
}

void removeByValue(Node *head, int n) {
	Node *prev = head;
	while(prev != NULL && prev->next->value != n) {
		prev = prev->next;
	}
	Node *tmp = prev->next;
	prev->next = tmp->next;
	free(tmp);
}

Node* append(Node *head, int n) {
	Node *new_node;
	new_node = malloc(sizeof(Node));
	new_node->value = n;
	Node *last = head;
	while (last->next != NULL) {
		last = last->next;
	}
	last->next = new_node;
	return new_node;
}

// int main() {
// 	Node *linkedList = LinkedList(7);
// 	append(linkedList, 19);
// 	append(linkedList, 18);
// 	append(linkedList, 17);
// 	append(linkedList, 16);
// 	append(linkedList, 15);
	
// 	print(linkedList);
// 	removeByValue(linkedList, 17);
// 	Node *t = search(linkedList, 17);
// 	if (t == NULL) {
// 		printf("not found\n");
// 	} else {
// 		printf("found %d\n", t->value);
// 	}
	
// 	return 0;
// }
